File: go/secret-handshake/secret_handshake.go
package secret

const (
	wink = 1 << iota
	blink
	closed
	jump
	reverse
)

func Handshake(code int) []string {
	a := make([]string, 0)
	if code < 0 {
		return a
	}

	a = addAction(code, wink, "wink", a)
	a = addAction(code, blink, "double blink", a)
	a = addAction(code, closed, "close your eyes", a)
	a = addAction(code, jump, "jump", a)

	if code&reverse == reverse {
		a = reversed(a)
	}
	return a
}

func addAction(code, flag int, action string, list []string) []string {
	if code&flag == flag {
		return append(list, action)
	}
	return list
}

func reversed(actions []string) []string {
	for i, j := 0, len(actions)-1; i < j; i, j = i+1, j-1 {
		actions[i], actions[j] = actions[j], actions[i]
	}
	return actions
}

File: elixir/secret-handshake/secret_handshake.exs
defmodule SecretHandshake do
  use Bitwise

  @command_flags %{
    1 => "wink",
    2 => "double blink",
    4 => "close your eyes",
    8 => "jump"
  }
  @reverse 16

  @doc """
  Determine the actions of a secret handshake based on the binary
  representation of the given `code`.

  If the following bits are set, include the corresponding action in your list
  of commands, in order from lowest to highest.

  1 = wink
  10 = double blink
  100 = close your eyes
  1000 = jump

  10000 = Reverse the order of the operations in the secret handshake
  """
  @spec commands(code :: integer) :: list(String.t())
  def commands(code) do
    is_reverse = band(code, @reverse) == @reverse
    actions = Enum.reduce(@command_flags, [],
                          SecretHandshake.reducer(code, is_reverse))
  end

  def reducer(code, reverse) when reverse do
    fn (x, xs) -> SecretHandshake.add_action(:prepend, code, x, xs) end
  end

  def reducer(code, reverse) do
    fn (x, xs) -> SecretHandshake.add_action(:append, code, x, xs) end
  end

  def add_action(:prepend, code, {flag, action}, acc) when band(code, flag) == flag do
    [action] ++ acc
  end

  def add_action(:append, code, {flag, action}, acc) when band(code, flag) == flag do
    acc ++ [action]
  end

  def add_action(_, _, _, acc) do
    acc
  end
end

